torch/package/package_exporter_no_torch.py

```diff
@@ -0,0 +1,1111 @@
+                ``allow_empty=False``, and :meth:`close` is called (either explicitly or via ``__exit__``) and the mock has
+                not been matched to a module used by the package being exported, an exception is thrown.
+                If ``allow_empty=True``, no such exception is thrown.
+
+        """
+        self.patterns[GlobGroup(include, exclude=exclude)] = _PatternInfo(
+            _ModuleProviderAction.MOCK, allow_empty
+        )
+
+    def extern(
+        self,
+        include: "GlobPattern",
+        *,
+        exclude: "GlobPattern" = (),
+        allow_empty: bool = True,
+    ):
+        """Include ``module`` in the list of external modules the package can import.
+        This will prevent dependency discovery from saving
+        it in the package. The importer will load an external module directly from the standard import system.
+        Code for extern modules must also exist in the process loading the package.
+
+        Args:
+            include (Union[List[str], str]): A string e.g. ``"my_package.my_subpackage"``, or list of strings
+                for the names of the modules to be externed. This can also be a glob-style pattern, as
+                described in :meth:`mock`.
+
+            exclude (Union[List[str], str]): An optional pattern that excludes some patterns that match the
+                include string.
+
+            allow_empty (bool): An optional flag that specifies whether the extern modules specified by this call
+                to the ``extern`` method must be matched to some module during packaging. If an extern module glob
+                pattern is added with ``allow_empty=False``, and :meth:`close` is called (either explicitly or via
+                ``__exit__``) before any modules match that pattern, an exception is thrown. If ``allow_empty=True``,
+                no such exception is thrown.
+
+        """
+        self.patterns[GlobGroup(include, exclude=exclude)] = _PatternInfo(
+            _ModuleProviderAction.EXTERN, allow_empty
+        )
+
+    def deny(self, include: "GlobPattern", *, exclude: "GlobPattern" = ()):
+        """Blocklist modules who names match the given glob patterns from the list of modules the package can import.
+        If a dependency on any matching packages is found, a :class:`PackagingError` is raised.
+
+        Args:
+            include (Union[List[str], str]): A string e.g. ``"my_package.my_subpackage"``, or list of strings
+                for the names of the modules to be externed. This can also be a glob-style pattern, as described in :meth:`mock`.
+
+            exclude (Union[List[str], str]): An optional pattern that excludes some patterns that match the include string.
+        """
+        self.patterns[GlobGroup(include, exclude=exclude)] = _PatternInfo(
+            _ModuleProviderAction.DENY, allow_empty=True
+        )
+
+    def persistent_id(self, obj):
```

Comment:
Yeah! Though I can refactor it to just move it into save_pickle.

Reply:
Scratch that .... I forgot that torch.package (non torch bits) is basically a wrapper around pickle. We can't just dump it into save_pickle. I clean it up a bit with the pattern you mentioned

Reply:
Scratch that .... I forgot that torch.package (non torch bits) is basically a wrapper around pickle. We can't just dump it into save_pickle. I clean it up a bit with the pattern you mentioned

Reply:
Sorry for the email spam. Since we want __reduce_pacakge__ support, and user persistent_id functionality should override ours. I think this is fine. This is also more or less the pattern pickle.py has

persistent_id - https://github.com/python/cpython/blob/082d3495d0c820972f09f6109a98ed7eb5a7b79f/Lib/pickle.py#L540

__reduce__ - https://github.com/python/cpython/blob/082d3495d0c820972f09f6109a98ed7eb5a7b79f/Lib/pickle.py#L580

Reply:
Sorry for the email spam. Since we want __reduce_pacakge__ support, and user persistent_id functionality should override ours. I think this is fine. This is also more or less the pattern pickle.py has

persistent_id - https://github.com/python/cpython/blob/082d3495d0c820972f09f6109a98ed7eb5a7b79f/Lib/pickle.py#L540

__reduce__ - https://github.com/python/cpython/blob/082d3495d0c820972f09f6109a98ed7eb5a7b79f/Lib/pickle.py#L580